source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc

```diff
@@ -18,7 +18,69 @@ namespace Extensions {
 namespace HttpFilters {
 namespace HeaderToMetadataFilter {
 
-Rule::Rule(const std::string& header, const ProtoRule& rule) : header_(header), rule_(rule) {
+// extract the value of the header
+absl::optional<std::string> HeaderValueSelector::extract(Http::HeaderMap& map) const {
+  const Http::HeaderEntry* header_entry = map.get(header_);
+  if (header_entry == nullptr) {
+    return absl::nullopt; // no value
+  }
+  // catch the value in the header
+  absl::optional<std::string> value = std::string(header_entry->value().getStringView());
+  if (remove_) {
+    map.remove(header_); // remove the header
+  }
+  return value;
+}
+
+// extract the value of the key from the cookie header
+absl::optional<std::string> CookieValueSelector::extract(Http::HeaderMap& map) const {
+  std::string value = Envoy::Http::Utility::parseCookieValue(map, cookie_);
+  if (!value.empty()) {
+    return absl::optional<std::string>(value);
+  }
+  return absl::nullopt; // no value
+}
+
+Rule::Rule(const ProtoRule& rule) : rule_(rule) {
+  // Ensure only one of header and cookie is specified.
+  // TODO(radha13): remove this once we are on v4 and these fields are folded into a oneof.
+  if (!rule.cookie().empty() && !rule.header().empty()) {
+    throw EnvoyException("Cannot specify both header and cookie");
+  }
+
+  // initialize the shared pointer
+  if (!rule.header().empty()) {
+    selector_ = std::make_shared<HeaderValueSelector>(Http::LowerCaseString(rule.header()), rule.remove());
+  } else if (!rule.cookie().empty()) {
+    selector_ = std::make_shared<CookieValueSelector>(rule.cookie());
+  } else {
+    throw EnvoyException("One of Cookie or Header option needs to be specified");
+  }
```

Comment:
this else block is redundant.... might as well remove it.

Reply:
I'm not sure. we used to have a test that checked whether a header was specified (I removed it but I plan to add it back). If we remove this the test will fail. Which means that we are not ensuring either one of header or cookie must be specified.

Reply:
Ah -- I see. Ok.